### PR TITLE
Use NanoGPT instead of MinGPT for training

### PR DESCRIPTION
**Pull request by igorjakus**
Making the training more efficient:
- **less training time** (difference even greater for longer runs, due to compilation overhead)
- **less GPU usage** (mine was like 98% in first run and 70% in second for fixed batch size)
- **lower loss** (not sure why, maybe better learning rate scheduling)

The code in `model.py` is like 99% copy-paste from [NanoGPT repository](https://github.com/karpathy/nanoGPT). 
It uses:
- FlashAttention (optimized CUDA kernels for Attention mechanism)
- mixed precision training (less VRAM usage)
- PyTorch GELU instead of self-written one
- optimized inference (only logits for last token, instead of logits for all tokens)
- Fused AdamW (optimized CUDA kernels)
- asynchroning, non-blocking transfer CPU->GPU
- [Cosine Annealing LR schedule](https://docs.pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html)
- Weight tying (`self.transformer.wte.weight = self.lm_head.weight`)
- No bias (probably because layernorm neutralizes it)

>>>>>>>>>> BEFORE CHANGE
```
(krasnal) ➜  krasnal git:(main) ✗ time python src/train.py
Loading dataset...
Vocab size: 1971
Dataset size: 79911 games
number of parameters: 11.60M
Using device: cuda
Starting training...
train (~4.00 ep): 100%|██████| 10000/10000 [05:26<00:00, 30.63iter/s, epoch=4.00, loss=2.9735]
Training finished.
Model saved to models/chess_model.pt and models/final_L6_H6_E384_I10000_B32_20260226_1953.pt
python src/train.py  323,68s user 6,55s system 99% cpu 5:30,40 total
(krasnal) ➜  krasnal git:(main) ✗
```

>>>>>>>>>>> AFTER CHANGE
```
(krasnal) ➜  krasnal git:(main) ✗ time python src/train.py
Loading dataset...
Vocab size: 1971
Dataset size: 79911 games
number of parameters: 11.38M
Using device: cuda
Mixed precision: bfloat16
num decayed parameter tensors: 26, with 11,570,304 parameters
num non-decayed parameter tensors: 13, with 4,992 parameters
using fused AdamW: True
Compiling model with torch.compile()...
Starting training...
train (~4.00 ep):   0%|                                                                                          | 8/10000 [02:11<42:28:04, 15.30s/iter, epoch=0.00, loss=7.6588, lr=0.00e+00]W0226 16:43:33.637000 20088 torch/_dynamo/convert_frame.py:1676] [0/8] torch._dynamo hit config.recompile_limit (8)
W0226 16:43:33.637000 20088 torch/_dynamo/convert_frame.py:1676] [0/8]    function: 'forward' (/home/igorjakus/Projects/krasnal/src/model.py:166)
W0226 16:43:33.637000 20088 torch/_dynamo/convert_frame.py:1676] [0/8]    last reason: 0/7: tensor 'idx' size mismatch at index 1. expected 149, actual 146
W0226 16:43:33.637000 20088 torch/_dynamo/convert_frame.py:1676] [0/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
W0226 16:43:33.637000 20088 torch/_dynamo/convert_frame.py:1676] [0/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/compile/programming_model.recompilation.html
train (~4.00 ep):  58%|████████████████████████████████████████████████████▋                                     | 5850/10000 [03:47<01:00, 68.41iter/s, epoch=2.34, loss=2.5343, lr=2.18e-0train (~4.00 ep):  59%|▌| 5854/10000 [03:47<01:00, 68.53iter/s, epoch=2.34, loss=2.5343, lr=2.                                                                                              train (~4.00 ep):  65%|▋| 6532/10000 [03:58<00:55, 62.98iter/s, epoch=2.61, loss=2.6674, lr=1.
train (~4.00 ep): 100%|█| 10000/10000 [04:57<00:00, 33.60iter/s, epoch=4.00, loss=2.3418, lr=5
Training finished.
Model saved to models/chess_model.pt and models/final_L6_H6_E384_I10000_B32_20260226_1646.pt
python src/train.py  294,29s user 28,63s system 105% cpu 5:05,66 total
```

### CODE REVIEW

File: src/config.py [hunk 1]
```diff
@@ -12,18 +12,27 @@
...
+    min_lr: float = 5e-5  # cosine annealing minimum LR
+    max_iters: int = 50000
+    warmup_iters: int = 100
     batch_size: int = 32
     num_workers: int = 4
+    weight_decay: float = 0.1
+    beta1: float = 0.9
+    beta2: float = 0.95
+    grad_clip: float = 1.0
+    compile: bool = True  # torch.compile: enable for long runs (10k+ iters)
```
[the remark this answers is not in the document]
Comment (aszkiel71) [reply src/config.py]: copilot halucyni

File: src/dataset.py [hunk 2]
```diff
@@ -1,21 +1,30 @@
...
+        self.dataset = HFDataset.from_parquet(str(parquet_path))
+        self.dataset.set_format(type="torch", columns=["token_ids"])
 
     def __len__(self):
-        return len(self.games)
+        return len(self.dataset)
 
     def __getitem__(self, idx: int):
-        return torch.tensor(self.games[idx], dtype=torch.long)
+        return self.dataset[idx]["token_ids"]
```
Comment (aszkiel71) [inline src/dataset.py]: return self.dataset[idx]["token_ids"].to(torch.long)

File: pyproject.toml [hunk 3]
```diff
@@ -26,5 +25,7 @@ dependencies = [
 [tool.ruff]
 line-length = 100
 
-[tool.uv.sources]
-mingpt = { git = "https://github.com/karpathy/minGPT.git" }
+[dependency-groups]
+dev = [
+    "pre-commit>=4.5.1",
+]
```
[the remark this answers is not in the document]
Comment (aszkiel71) [reply pyproject.toml]: copilot dobrze mowi.

File: src/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (igorjakus) [reply src/config.py]: imo nie, docelowo w treningu chcielibysmy robic compile = True, on Ci robi w giga duzo sztuczek typu fuzje kerneli CUDA, tracimy overhead pythona, mniejsze zuzycie pamieci itp.

File: src/train.py [hunk 4]
```diff
@@ -17,76 +47,127 @@ def main():
...
+    )
 
-    print(f"Using device: {train_config.device}")
+    model.to(device)
 
-    optimizer = model.configure_optimizers(train_config)
+    # torch.compile
+    if tconf.compile and device_type == "cuda":
+        print("Compiling model with torch.compile()...")
+        model = torch.compile(model, dynamic=True)  # dynamic padding
```
[the remark this answers is not in the document]
Comment (igorjakus) [reply src/train.py]: giga madre, czyli np. minimalny padding 16, potem 32, 64, 128, 256, 512, 1024 i bedzie tylko 7 recompile wiec z zapasem
